apps/ssr-tests-v9/src/test.ts

```diff
@@ -0,0 +1,93 @@
+import chalk from 'chalk';
+import * as fs from 'fs';
+import * as path from 'path';
+import { Browser } from 'puppeteer';
+
+import { PROVIDER_ID } from './utils/constants';
+import { hrToSeconds } from './utils/helpers';
+import { launchBrowser } from './utils/launchBrowser';
+
+class RenderError extends Error {}
+
+export async function runTest(browser: Browser, url: string): Promise<void> {
+  const page = await browser.newPage();
+  await page.setRequestInterception(true);
+
+  let error: Error | undefined;
+
+  page.on('console', message => {
+    if (message.type() === 'error') {
+      // Ignoring network errors as we have an interceptor that prevents loading everything except our JS bundle
+      if (!message.text().includes('net::ERR_FAILED')) {
+        error = new RenderError(message.text());
+      }
+    }
+  });
+
+  page.on('request', request => {
+    // Our interceptor allows only our HTML and JS output
+    if (request.url() === url || request.url().endsWith('/out-esm.js')) {
+      return request.continue();
+    }
+
+    return request.abort();
+  });
+
+  page.on('pageerror', err => {
+    error = err;
+  });
+
+  await page.goto(url);
+  await page.waitForSelector(`#${PROVIDER_ID}`);
+
+  await page.close();
+
+  if (error) {
+    throw error;
+  }
+}
+
+async function test(): Promise<void> {
+  const startTime = process.hrtime();
+  console.log('Starting a browser...');
+
+  const browser = await launchBrowser();
+  console.log('Using', await browser.version());
+
+  const htmlPath = path.resolve(__dirname, '..', 'dist', 'index.html');
+
+  if (!fs.existsSync(htmlPath)) {
+    throw new Error('"dist/index.html" does not exist, please run "yarn build" first');
```

Comment:
**thought:** for better DX, can we maybe initiate the build for the user with proper notification what's going on and re-run the test ?